ingest/src/main/java/com/microsoft/azure/kusto/ingest/ResourceAlgorithms.java

```diff
@@ -0,0 +1,136 @@
+    }
+
+    public static String uploadStreamToBlobWithRetries(ResourceManager resourceManager, AzureStorageClient azureStorageClient, InputStream stream,
+            String blobName, boolean shouldCompress)
+            throws IngestionClientException, IngestionServiceException {
+        return resourceActionWithRetries(resourceManager, resourceManager.getShuffledContainers(), container -> {
+            azureStorageClient.uploadStreamToBlob(stream, blobName, container.getContainer(), shouldCompress);
+            return (container.getContainer().getBlobContainerUrl() + "/" + blobName + container.getSas());
+        }, "ResourceAlgorithms.uploadLocalFileWithRetries", Collections.emptyMap());
+    }
+
+    public static String uploadLocalFileWithRetries(ResourceManager resourceManager, AzureStorageClient azureStorageClient, File file, String blobName,
+            boolean shouldCompress)
+            throws IngestionClientException, IngestionServiceException {
+        return resourceActionWithRetries(resourceManager, resourceManager.getShuffledContainers(), container -> {
+            azureStorageClient.uploadLocalFileToBlob(file, blobName, container.getContainer(), shouldCompress);
+            return (container.getContainer().getBlobContainerUrl() + "/" + blobName + container.getSas());
+        }, "ResourceAlgorithms.uploadLocalFileWithRetries", Collections.emptyMap());
+    }
+
+    @NotNull
+    public static <T> List<T> roundRobinNestedList(@NotNull List<List<T>> validResources) {
+        int longestResourceList = validResources.stream().mapToInt(List::size).max().orElse(0);
+
+        // Go from 0 to the longest list length
+        return IntStream.range(0, longestResourceList).boxed()
+                // This flat maps combines all the inner lists
+                .flatMap(i ->
+                // For each list, get the i'th element if it exists, or null otherwise (if the list is shorter)
+                validResources.stream().map(r -> r.size() > i ? r.get(i) : null)
+                        // Remove nulls
+                        .filter(Objects::nonNull))
+                // So we combine the list of the first element of each list, then the second element, etc.
+                .collect(Collectors.toList());
+    }
+
+    public static <T extends ResourceWithSas<?>> List<T> getShuffledResources(List<RankedStorageAccount> shuffledAccounts, List<T> resourceOfType) {
+        Map<String, List<T>> accountToResourcesMap = groupResourceByAccountName(resourceOfType);
+
+        List<List<T>> validResources = shuffledAccounts.stream()
+                // For each shuffled account, get the resources for that account
+                .map(account -> accountToResourcesMap.get(account.getAccountName()))
+                // Remove nulls and empty lists
+                .filter(resourceList -> resourceList != null && !resourceList.isEmpty())
```

Comment:
These are all of the accounts we have, and we select one type of resource from them - so if we select queues, we could have an account which only has containers, so it will be empty. Null I don't think they can, but worth checking if we're already checking empty.